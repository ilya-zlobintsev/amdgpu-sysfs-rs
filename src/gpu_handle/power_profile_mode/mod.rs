//! `pp-power-profile-mode`
mod basic;
mod full;

use crate::Result;
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

pub use basic::BasicTable;
pub use full::{FullTable, FullTableMode};

/// Table of predefined power profile modes with a list of GPU-specific heuristics

/// https://kernel.org/doc/html/latest/gpu/amdgpu/thermal.html#pp-power-profile-mode
#[derive(Debug)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(
    feature = "serde",
    serde(tag = "kind", content = "data", rename_all = "snake_case")
)]
pub enum PowerProfileModesTable {
    /// The full table format, used by dedicated GPUs
    Full(FullTable),
    /// Basic table format
    Basic(BasicTable),
}

impl PowerProfileModesTable {
    /// Parse the table from a given string
    pub fn parse(s: &str) -> Result<Self> {
        // The basic format starts with a number
        if s.split_whitespace()
            .next()
            .and_then(|item| item.parse::<usize>().ok())
            .is_some()
        {
            BasicTable::parse(s).map(Self::Basic)
        } else {
            FullTable::parse(s).map(Self::Full)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::PowerProfileModesTable;
    use insta::assert_yaml_snapshot;

    const TABLE_VEGA56: &str = include_test_data!("vega56/pp_power_profile_mode");
    const TABLE_RX580: &str = include_test_data!("rx580/pp_power_profile_mode");
    const TABLE_4800H: &str = include_test_data!("internal-4800h/pp_power_profile_mode");

    #[test]
    fn parse_full_vega56() {
        let table = PowerProfileModesTable::parse(TABLE_VEGA56).unwrap();
        assert_yaml_snapshot!(table, {
            ".data.modes[].heuristics" => insta::sorted_redaction()
        });
    }

    #[test]
    fn parse_full_rx580() {
        let table = PowerProfileModesTable::parse(TABLE_RX580).unwrap();
        assert_yaml_snapshot!(table, {
            ".data.modes[].heuristics" => insta::sorted_redaction()
        });
    }

    #[test]
    fn parse_full_internal_4800h() {
        let table = PowerProfileModesTable::parse(TABLE_4800H).unwrap();
        assert_yaml_snapshot!(table);
    }
}
