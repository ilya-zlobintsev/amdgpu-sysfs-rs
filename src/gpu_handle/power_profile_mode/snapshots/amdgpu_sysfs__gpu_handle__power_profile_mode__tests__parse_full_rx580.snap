---
source: src/gpu_handle/power_profile_mode/mod.rs
expression: table
---
kind: full
data:
  modes:
    - name: BOOTUP_DEFAULT
      heuristics:
        MCLK_ACTIVE_LEVEL: ~
        MCLK_DOWN_HYST: ~
        MCLK_UP_HYST: ~
        SCLK_ACTIVE_LEVEL: ~
        SCLK_DOWN_HYST: ~
        SCLK_UP_HYST: ~
    - name: 3D_FULL_SCREEN
      heuristics:
        MCLK_ACTIVE_LEVEL: "25"
        MCLK_DOWN_HYST: "60"
        MCLK_UP_HYST: "10"
        SCLK_ACTIVE_LEVEL: "30"
        SCLK_DOWN_HYST: "100"
        SCLK_UP_HYST: "0"
    - name: POWER_SAVING
      heuristics:
        MCLK_ACTIVE_LEVEL: ~
        MCLK_DOWN_HYST: ~
        MCLK_UP_HYST: ~
        SCLK_ACTIVE_LEVEL: "30"
        SCLK_DOWN_HYST: "0"
        SCLK_UP_HYST: "10"
    - name: VIDEO
      heuristics:
        MCLK_ACTIVE_LEVEL: "31"
        MCLK_DOWN_HYST: "16"
        MCLK_UP_HYST: "10"
        SCLK_ACTIVE_LEVEL: ~
        SCLK_DOWN_HYST: ~
        SCLK_UP_HYST: ~
    - name: VR
      heuristics:
        MCLK_ACTIVE_LEVEL: "10"
        MCLK_DOWN_HYST: "100"
        MCLK_UP_HYST: "0"
        SCLK_ACTIVE_LEVEL: "50"
        SCLK_DOWN_HYST: "11"
        SCLK_UP_HYST: "0"
    - name: COMPUTE
      heuristics:
        MCLK_ACTIVE_LEVEL: ~
        MCLK_DOWN_HYST: ~
        MCLK_UP_HYST: ~
        SCLK_ACTIVE_LEVEL: "30"
        SCLK_DOWN_HYST: "5"
        SCLK_UP_HYST: "0"
    - name: CUSTOM
      heuristics:
        MCLK_ACTIVE_LEVEL: ~
        MCLK_DOWN_HYST: ~
        MCLK_UP_HYST: ~
        SCLK_ACTIVE_LEVEL: ~
        SCLK_DOWN_HYST: ~
        SCLK_UP_HYST: ~
  active: 1
  available_heuristics:
    - SCLK_UP_HYST
    - SCLK_DOWN_HYST
    - SCLK_ACTIVE_LEVEL
    - MCLK_UP_HYST
    - MCLK_DOWN_HYST
    - MCLK_ACTIVE_LEVEL

